CA_DataUploaderLib/IOconf/IOconfFilter.cs

```diff
@@ -26,20 +33,96 @@ public IOconfFilter(string row, int lineNum) : base(row, lineNum, "Filter")
+            }
+            catch (FormatException) { throw; }
+            catch (Exception ex)
+            {
+                throw new Exception($"Wrong format for filter expression: {row}{Environment.NewLine}{format}", ex);
+            }
         }
 
         public override IEnumerable<string> GetExpandedSensorNames(IIOconf ioconf)
         {
             yield return NameInVector;
         }
+
+        public IEnumerable<string> GetDecisionFields() => 
+            filterType == FilterType.Sustained 
+                ? [NameInVector, NameInVector + "_targettime"]
+                : throw new InvalidOperationException($"Unexpected call to GetDecisionFields for filter type {Row}");
+
+        public void MakeDecision(MathVectorExpansion.MathContext context)
+        {
+            if (filterType != FilterType.Sustained) 
+                throw new InvalidOperationException($"Unexpected MakeDecision call for non decision filter: {Row}");
+
+            MakeDecisionSustained(context);
+        }
+
+        private void MakeDecisionSustained(MathVectorExpansion.MathContext context)
+        {
+            var (time, vector) = (context.Vector.Timestamp, context.Vector.Data);
+            var expr = _sustainedExpression ?? throw new InvalidOperationException($"Unexpected null expression for filter: {Row}");
+            ref var targetTime = ref vector[_sustainedTargetTimeFieldIndex];
+            ref var output = ref vector[_decisionFieldIndex];
+
+            if (!context.CalculateBoolean(expr))
+                (output, targetTime) = (0, 0);//the condition is not met, reset both the filter output and target time
+            else if (targetTime == 0)
+                (output, targetTime) = (0, time.AddSeconds(filterLength).ToOADate());//the first cycle meeting the condition sets the target time in filterLength seconds
+            else
+                output = DateTime.FromOADate(targetTime) >= time ? 1 : 0; //set the output to 1 if we already reached the target time
+        }
+
+        public void Initialize(List<string> fields)
+        {
+            if (filterType != FilterType.Sustained) 
+                throw new InvalidOperationException($"Unexpected Initialize call for non decision filter: {Row}");
+
+            _decisionFieldIndex = fields.IndexOf(Name);
+            if (_decisionFieldIndex < 0) throw new ArgumentException($"{Name} was not found in received vector fields", nameof(fields));
+            _sustainedTargetTimeFieldIndex = fields.IndexOf(Name + "_targettime");
+            if (_sustainedTargetTimeFieldIndex < 0) throw new ArgumentException($"{Name} was not found in received vector fields", nameof(fields));
+            foreach (var source in SourceNames)
+            {
+                if (!fields.Contains(source))
+                    throw new ArgumentException($"Filter {Name} uses {source} which was not found in received vector fields", nameof(fields));
```

Comment:
in case there are more sources not found - how about listing them all in this message?